cmd/tsbs_load_clickhouse/creator.go

```diff
@@ -0,0 +1,238 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"strings"
+
+	"github.com/jmoiron/sqlx"
+)
+
+// loader.DBCreator interface implementation
+type dbCreator struct {
+	tags    string
+	cols    []string
+	connStr string
+}
+
+// loader.DBCreator interface implementation
+func (d *dbCreator) Init() {
+	br := loader.GetBufferedReader()
+	d.readDataHeader(br)
+}
+
+// readDataHeader fills dbCreator struct with data structure (tables description)
+// specified at the beginning of the data file
+func (d *dbCreator) readDataHeader(br *bufio.Reader) {
+	// First N lines are header, describing data structure.
+	// The first line containing tags table name ('tags') followed by list of tags, comma-separated.
+	// Ex.: tags,hostname,region,datacenter,rack,os,arch,team,service,service_version
+	// The second through N-1 line containing table name (ex.: 'cpu') followed by list of column names,
+	// comma-separated. Ex.: cpu,usage_user,usage_system,usage_idle,usage_nice,usage_iowait,usage_irq,usage_softirq
+	// The last line being blank to separate from the data
+	//
+	// Header example:
+	// tags,hostname,region,datacenter,rack,os,arch,team,service,service_version,service_environment
+	// cpu,usage_user,usage_system,usage_idle,usage_nice,usage_iowait,usage_irq,usage_softirq,usage_steal,usage_guest,usage_guest_nice
+	// disk,total,free,used,used_percent,inodes_total,inodes_free,inodes_used
+	// nginx,accepts,active,handled,reading,requests,waiting,writing
+
+	i := 0
+	for {
+		var err error
+		var line string
+
+		if i == 0 {
+			// read first line - list of tags
+			d.tags, err = br.ReadString('\n')
+			if err != nil {
+				fatal("input has wrong header format: %v", err)
+			}
+			d.tags = strings.TrimSpace(d.tags)
+		} else {
+			// read the second and further lines - metrics descriptions
+			line, err = br.ReadString('\n')
+			if err != nil {
+				fatal("input has wrong header format: %v", err)
+			}
+			line = strings.TrimSpace(line)
+			if len(line) == 0 {
+				// empty line - end of header
+				break
+			}
+			// append new table/columns set to the list of tables/columns set
+			d.cols = append(d.cols, line)
+		}
+		i++
+	}
+}
+
+// loader.DBCreator interface implementation
+func (d *dbCreator) DBExists(dbName string) bool {
+	db := sqlx.MustConnect(dbType, getConnectString(false))
+	defer db.Close()
+
+	sql := "SELECT name, engine FROM system.databases WHERE name = '" + dbName + "'"
```

Comment:
You mostly use string concatenation for dynamically generating strings in this package, whereas you use string formatting (eg fmt.Sprintf) in the query generation package. Any reason for this discrepancy? I generally find the concatenation harder to read.

Reply:
There is no strict rule, but main idea is that short one-line-strings are constructed using concatenation, while long-multi-line strings are fmt.Sprintf-ed. It is exactly like this in `creator.go` under discussion, where one-line SQL queries (like short `SELECT` or `CREATE DATABASE`) are concatenated and multi-line queries (like `CREATE TABLE long table spec goes here`) are fmt.Sprintf-ed. The threshold after which string must not be concatenated is, I believe, line break. However, this is just personal.

Reply:
Updated to `fmt.Sprintf`